### ISSUE TITLE
room for performance improvement for SubArray

### ISSUE THREAD

**Issue by alsam** [JuliaLang/julia#5117]
For the following synthetic testcase:

```julia

function mykern(r, s, a, x, y)
    const nsteps = 25000 
    istep = 0
    
    ifirst = minimum(r)
    ilast  = maximum(r)
    while istep < nsteps #&& t < tmax

        for i = ifirst : ilast-1
            s[i] = a * (x[i+1] - x[i]) / (y[i+1] - y[i])
        end

        for i = ifirst : ilast
            x[i] = a * s[i-1]
        end

        for i = ifirst : ilast
            y[i] = a * x[i-1]
        end

        for i = ifirst : ilast-1
            s[i] = s[i] - (x[i+1] - x[i]) / (y[i+1] - y[i])
        end
        istep=istep + 1
    end 
    s
end 

const siz = 10^4
const a = e^pi - pi^e

s = rand(siz)
x = rand(siz)
y = rand(siz)

s1 = sub(s, 1:(siz - 1))
x1 = sub(x, 1:(siz - 1))
y1 = sub(y, 1:(siz - 1))

s2 = sub(s, 2:siz)
x2 = sub(x, 2:siz)
y2 = sub(y, 2:siz)

s3 = sub(s, 4:siz)

# warming up
mykern( 2:(siz - 3), s, a, x,  y  )
mykern( 2:(siz - 3), s, a, x1, y1 )

# the real test

println("follow up: mykern( 2:(siz - 3), s, a, x, y )")
@time mykern( 2:(siz - 3), s, a, x,  y  )
println("follow up: mykern( 3:(siz - 2), s2, a, x2, y2 )")
@time mykern( 3:(siz - 2), s2, a, x2, y2 )

println("profiling: mykern( 2:(siz - 3), s, a, x, y )")
@profile mykern( 2:(siz - 3), s, a, x,  y  )
Profile.print()
#code_typed(mykern,(2:(siz - 3), s, a, x,  y))
code_llvm(mykern,(2:(siz - 3), s, a, x,  y))
code_native(mykern,(2:(siz - 3), s, a, x,  y))

println("profiling: mykern( 3:(siz - 2), s2, a, x2, y2 )")
@profile mykern( 3:(siz - 2), s2, a, x2, y2 )
Profile.print()
#code_typed(mykern,(3:(siz - 2), s2, a, x2, y2))
code_llvm(mykern,(3:(siz - 2), s2, a, x2, y2))
code_native(mykern,(3:(siz - 2), s2, a, x2, y2))
```

SubArray based arrays got significant - up to 4x performance slowdown than Array based ones.

Profiling diff:

```
-follow up: mykern( 2:(siz - 3), s, a, x, y )
-elapsed time: 2.848171233 seconds (611520 bytes allocated)
-follow up: mykern( 3:(siz - 2), s2, a, x2, y2 )
-elapsed time: 8.839612056 seconds (704324 bytes allocated)
-profiling: mykern( 2:(siz - 3), s, a, x, y )
-2807 boot.jl; include; line: 238
-    2807 profile.jl; anonymous; line: 14
-       98  ...ray/src/shifted3.jl; mykern; line: 10
-       801 ...ray/src/shifted3.jl; mykern; line: 11
-       149 ...ray/src/shifted3.jl; mykern; line: 14
-       344 ...ray/src/shifted3.jl; mykern; line: 15
-       96  ...ray/src/shifted3.jl; mykern; line: 18
-       430 ...ray/src/shifted3.jl; mykern; line: 19
-       889 ...ray/src/shifted3.jl; mykern; line: 23
+profiling: mykern( 3:(siz - 2), s2, a, x2, y2 )
+11591 boot.jl; include; line: 238
+    11591 profile.jl; anonymous; line: 14
+       185  ...ray/src/shifted3.jl; mykern; line: 10
+       3302 ...ray/src/shifted3.jl; mykern; line: 11
+       149  ...ray/src/shifted3.jl; mykern; line: 14
+       2078 ...ray/src/shifted3.jl; mykern; line: 15
+       96   ...ray/src/shifted3.jl; mykern; line: 18
+       2133 ...ray/src/shifted3.jl; mykern; line: 19
+       3648 ...ray/src/shifted3.jl; mykern; line: 23
```

revealed that for SubArray code was invoked more frequently, say for the line 11
```julia
            s[i] = a * (x[i+1] - x[i]) / (y[i+1] - y[i])
```

The Array version was invoked 801 times, SubArray version - 3301, i.e. 4x time more frequently.

Any hints for more detailed profiling/study of the code?

**Comment by alsam** [comment]
Thanks Tim for your efforts.
Just noticed and got an occasion to measure it again.
Very strange, my measurements show huge performance gap, otherwise I'd rather switch to standard SubArrays way as 40% penalty looks attractive. The code
```julia
function mykern(r, s, a, x, y)
    const nsteps = 250 
    istep = 0

    ifirst = minimum(r)
    ilast  = maximum(r)
    while istep < nsteps #&& t < tmax

        @inbounds for i = ifirst : ilast-1
            s[i] = a * (x[i+1] - x[i]) / (y[i+1] - y[i]) # line #10
        end

        @inbounds for i = ifirst : ilast
            x[i] = a * s[i-1]                            # line #14
        end

        @inbounds for i = ifirst : ilast
            y[i] = a * x[i-1]
        end

        @inbounds for i = ifirst : ilast-1
            s[i] = s[i] - (x[i+1] - x[i]) / (y[i+1] - y[i])
        end
        istep=istep + 1
    end 
    s
end 

const siz = 10^4
const a = e^pi - pi^e

s = rand(siz)
x = rand(siz)
y = rand(siz)

s1 = sub(s, 1:(siz - 1))
x1 = sub(x, 1:(siz - 1))
y1 = sub(y, 1:(siz - 1))

s2 = sub(s, 2:siz)
x2 = sub(x, 2:siz)
y2 = sub(y, 2:siz)

s3 = sub(s, 4:siz)

# warming up
mykern( 2:(siz - 3), s, a, x,  y  )
mykern( 2:(siz - 3), s, a, x1, y1 )
mykern( 3:(siz - 2), s, a, x2, y2 )
# @time mykern( 2:(siz - 3), s, a, x,  y )

# the real test

println("follow up: mykern( 2:(siz - 3), s, a, x, y )")
@time mykern( 2:(siz - 3), s, a, x,  y  )
println("follow up: mykern( 2:(siz - 3), s1, a, x1, y1 )")
@time mykern( 2:(siz - 3), s1, a, x1, y1 )
println("follow up: mykern( 3:(siz - 2), s2, a, x2, y2 )")
@time mykern( 3:(siz - 2), s2, a, x2, y2 )
@time mykern( 3:(siz - 2), s1, a, x2, y2 )
@time mykern( 5:(siz - 5), s3, a, x2, y2 )
```
```sh
julia shifted3.jl 
follow up: mykern( 2:(siz - 3), s, a, x, y )
  17.038 milliseconds (165 allocations: 11149 bytes)
follow up: mykern( 2:(siz - 3), s1, a, x1, y1 )
   9.040 seconds      (73071 k allocations: 1115 MB, 1.11% gc time)
follow up: mykern( 3:(siz - 2), s2, a, x2, y2 )
   8.964 seconds      (73056 k allocations: 1115 MB, 0.94% gc time)
   8.449 seconds      (73056 k allocations: 1115 MB, 0.88% gc time)
   8.462 seconds      (73026 k allocations: 1114 MB, 0.89% gc time)
```
What's wrong with the code above?
Thanks!

**Comment by timholy** [comment]
My results:
```
follow up: mykern( 2:(siz - 3), s, a, x, y )
  11.795 milliseconds (6 allocations: 192 bytes)
follow up: mykern( 2:(siz - 3), s1, a, x1, y1 )
 108.014 milliseconds (38263 allocations: 1119 KB)
follow up: mykern( 3:(siz - 2), s2, a, x2, y2 )
  16.299 milliseconds (6 allocations: 192 bytes)
  16.692 milliseconds (6 allocations: 192 bytes)
  22.398 milliseconds (6 allocations: 192 bytes)
```
Are you by any chance running julia 0.3? This is all only only julia 0.4.

**Comment by ScottPJones** [comment]
That has to be 0.4, unless somebody backported my `@time` macro changes...

**Comment by timholy** [comment]
Yeah, good point. What is your julia version, anyway?

**Comment by ScottPJones** [comment]
Also, with that large number of allocations (~73 million, ~1GB), it doesn't surprise me that it's slow...
I'd look into that first off.

**Comment by alsam** [comment]
Your timing is good. I wish I had the same.
My julia version:
```julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "help()" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.0-dev+5208 (2015-06-04 23:18 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit e9fa25b* (0 days old master)
|__/                   |  x86_64-linux-gnu
```

**Comment by timholy** [comment]
That's why I assumed julia 0.3, since SubArray construction was type-unstable on 0.3.

Also @alsam, your "warmup" doesn't really warm up the case where a subarray is the 2nd argument of `mykern`. So it _could_ be computation, but if so there's still something wrong with your julia (compilation shouldn't be that slow). But do try
```jl
include("shifted3.jl")
include("shifted3.jl")
```
in the REPL and see if the 2nd call is faster.

**Comment by alsam** [comment]
Just tried to 
```julia
include("shifted3.jl")
include("shifted3.jl")
```
The performance is also bad unfortunately. The experiment with rotating parameters to `mykern` in order to make subarray the 1st argument didn't improve performance as well. Maybe something wrong with julia installation. Will retry with nightly binary and at home (another installation). Thanks!

**Comment by timholy** [comment]
Yeah, it sounds like you need to try a different julia.

FYI, `@profile` is oh-so-much more informative than `@time`, whenever you want to know details about performance. Might be handy if switching julia sources doesn't work out.

**Comment by alsam** [comment]
Ok, this is just a regression. Nightly binary has the same version and suffers from the same performance degradation.
With 0.3.9 it is ok:
```julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "help()" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.3.9 (2015-05-30 11:24 UTC)
 _/ |\__'_|_|_|\__'_|  |  Official http://julialang.org/ release
|__/                   |  x86_64-unknown-linux-gnu
```
```sh
julia shifted3.jl 
follow up: mykern( 2:(siz - 3), s, a, x, y )
elapsed time: 0.013820144 seconds (13888 bytes allocated)
follow up: mykern( 2:(siz - 3), s1, a, x1, y1 )
elapsed time: 0.156636415 seconds (938696 bytes allocated)
follow up: mykern( 3:(siz - 2), s2, a, x2, y2 )
elapsed time: 0.109257333 seconds (88 bytes allocated)
elapsed time: 0.109363798 seconds (88 bytes allocated)
elapsed time: 0.110168689 seconds (88 bytes allocated)
```
Thanks

**Comment by timholy** [comment]
Note the SubArray performance is ~10x worse on 0.3---it's only on 0.4 that you have a hope of getting something close to the performance of regular Arrays. 

That this is broken with the binaries but not source-builds speaks to a serious problem with the binaries. Can you profile on 0.4? (See http://docs.julialang.org/en/latest/manual/profile/). If you want to use SubArrays for performance reasons, you---or someone---will need to figure out why the binaries are so borked.

CC @staticfloat.

**Comment by alsam** [comment]
Ok, here it is, for nightly slow version
```julia
julia-nightly 
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "help()" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.0-dev+5208 (2015-06-04 23:18 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit e9fa25b* (0 days old master)
|__/                   |  x86_64-unknown-linux-gnu
```
```sh
julia-nightly shifted3.jl 
follow up: mykern( 2:(siz - 3), s, a, x, y )
  12.965 milliseconds (167 allocations: 12813 bytes)
follow up: mykern( 2:(siz - 3), s1, a, x1, y1 )
   7.799 seconds      (73071 k allocations: 1115 MB, 1.25% gc time)
follow up: mykern( 3:(siz - 2), s2, a, x2, y2 )
   8.685 seconds      (73056 k allocations: 1115 MB, 1.50% gc time)
   7.972 seconds      (73056 k allocations: 1115 MB, 0.97% gc time)
   7.900 seconds      (73026 k allocations: 1114 MB, 0.97% gc time)
8630 ./client.jl; _start; line: 412
 8630 ./client.jl; process_options; line: 320
  8630 loading.jl; include_from_node1; line: 133
   8630 ./boot.jl; include; line: 253
    8630 profile.jl; anonymous; line: 16
     2874 ...ROGRAM0/shifted3.jl; mykern; line: 10
      2565 abstractarray.jl; checkbounds; line: 160
      65   abstractarray.jl; checkbounds; line: 166
       20 abstractarray.jl; trailingsize; line: 99
       23 abstractarray.jl; trailingsize; line: 100
      7    abstractarray.jl; trailingsize; line: 99
     3    ...ROGRAM0/shifted3.jl; mykern; line: 13
     1149 ...ROGRAM0/shifted3.jl; mykern; line: 14
      1035 abstractarray.jl; checkbounds; line: 160
      35   abstractarray.jl; checkbounds; line: 166
       14 abstractarray.jl; trailingsize; line: 99
       10 abstractarray.jl; trailingsize; line: 100
      4    abstractarray.jl; trailingsize; line: 99
     1151 ...ROGRAM0/shifted3.jl; mykern; line: 18
      1027 abstractarray.jl; checkbounds; line: 160
      27   abstractarray.jl; checkbounds; line: 166
       4  abstractarray.jl; trailingsize; line: 99
       13 abstractarray.jl; trailingsize; line: 100
      5    abstractarray.jl; trailingsize; line: 99
     3425 ...ROGRAM0/shifted3.jl; mykern; line: 22
      3062 abstractarray.jl; checkbounds; line: 160
      89   abstractarray.jl; checkbounds; line: 166
       23 abstractarray.jl; trailingsize; line: 99
       31 abstractarray.jl; trailingsize; line: 100
      4    abstractarray.jl; trailingsize; line: 99
     16   abstractarray.jl; checkbounds; line: 160
```
For the 0.3.9 one:
```sh
julia shifted3.jl 
follow up: mykern( 2:(siz - 3), s, a, x, y )
elapsed time: 0.013353391 seconds (13888 bytes allocated)
follow up: mykern( 2:(siz - 3), s1, a, x1, y1 )
elapsed time: 0.157818713 seconds (938696 bytes allocated)
follow up: mykern( 3:(siz - 2), s2, a, x2, y2 )
elapsed time: 0.112287131 seconds (88 bytes allocated)
elapsed time: 0.112516582 seconds (88 bytes allocated)
elapsed time: 0.112086062 seconds (88 bytes allocated)
117 ./client.jl; _start; line: 354
 117 ./client.jl; process_options; line: 285
  117 loading.jl; include_from_node1; line: 128
   117 ./boot.jl; include; line: 245
    117 profile.jl; anonymous; line: 14
     34 .../PROGRAM0/shifted3.jl; mykern; line: 10
     2  .../PROGRAM0/shifted3.jl; mykern; line: 13
     17 .../PROGRAM0/shifted3.jl; mykern; line: 14
     20 .../PROGRAM0/shifted3.jl; mykern; line: 18
     44 .../PROGRAM0/shifted3.jl; mykern; line: 22
```
I just added:
```julia
Profile.clear()
@profile mykern( 2:(siz - 3), s1, a, x1, y1 )
Profile.print()
```
to the code above.
As I see nightly does a lot of bounds checking inside abstractarray.jl

**Comment by KristofferC** [comment]
FWIW I can reproduce the slowdown with the nightly build but when compiling from source, it is fast.

**Comment by alsam** [comment]
Thanks for finding this.
BTW with 0.3.9 the following trick to emulate Fortran indexing works
```julia
julia> a = [-10:10];
julia> b = sub(a, 12:21);
julia> b[-11]
ERROR: BoundsError()
 in getindex at subarray.jl:194
julia> b[-10]
-10
julia> b[10]
10
```
but with nightly access to `b[-10]` triggers an exception due to `checkbounds`

**Comment by KristofferC** [comment]
Actually I am wrong, I did not have the latest source built. Right now I am getting the same with the nighly and the source build(!)

In both cases it seems that it produces a `checkbounds` even with explicitly running with `check-bounds=no` and the `checkbounds` is also type unstable:

```julia
julia> @code_warntype getindex(s1, 1)
Variables:
  V::SubArray{Float64,1,Array{Float64,1},Tuple{UnitRange{Int64}},1}
  I::Tuple{Int64}

Body:
  begin 
      $(Expr(:meta, :inline)) # subarray2.jl, line 4:
      checkbounds(V::SubArray{Float64,1,Array{Float64,1},Tuple{UnitRange{Int64}},1},(top(getfield))(I::Tuple{Int64},1)::Int64)::Union(Bool,Void)
      return arrayref((top(getfield))(V::SubArray{Float64,1,Array{Float64,1},Tuple{UnitRange{Int64}},1},:parent)::Array{Float64,1},box(Int64,sub_int(box(Int64,add_int((top(getfield))(V::SubArray{Float64,1,Array{Float64,1},Tuple{UnitRange{Int64}},1},:first_index)::Int64,(top(getfield))(I::Tuple{Int64},1)::Int64)),1)))::Float64
  end::Float64
```

**Comment by mbauman** [comment]
It should be just fine that `checkbounds` is type-unstable since its output is never used.  In #10525, I added bounds checking to `subarray`, with measured performance drops of ~10-20%.  Something is drastically wrong if performance dropped by ~500x!

**Comment by KristofferC** [comment]
I will check before #10525 and see how the situation is.

**Comment by mbauman** [comment]
Yeah, that's because we don't have a solution to #7799 yet.  Instead of `@inbounds` you need to use `Base.unsafe_getindex`.

**Comment by tkelman** [comment]
One difference between nightly and source builds is nightlies limit LLVM to x86_64 instruction set, whereas source builds might be using avx etc. What kind of processor does `versioninfo()` say you have?

**Comment by KristofferC** [comment]
@tkelman This problem for me seem to occur both on source and nightly.

**Comment by KristofferC** [comment]
In the commit before #10525 got merged, the code runs fast.

Regarding "It should be just fine that checkbounds is type-unstable since its output is never used."

Before #10525
```julia
julia> @time mykern( 3:(siz - 2), s2, a, x2, y2 );
   8.013 milliseconds (6 allocations: 192 bytes)
```

After
```julia
julia> @time mykern( 3:(siz - 2), s2, a, x2, y2 );
   6.023 seconds      (73056 k allocations: 1115 MB, 0.69% gc time)
```

All time is spent in `checkbounds`.

**Comment by mbauman** [comment]
> It should be just fine that checkbounds is type-unstable since its output is never used.

Whoa.  Is my assumption there wrong in some cases!?  It certainly doesn't happen all the time.

Rewriting to use `unsafe_(set|get)index` makes all the spurious allocations go away.  And the only difference is `checkbounds`.

**Comment by tkelman** [comment]
Worth adding this benchmark to `test/perf` somewhere? Also probably good idea to take a careful look at perf comparisons before/after #10525.

**Comment by yuyichao** [comment]
> Whoa. Is my assumption there wrong in some cases!? It certainly doesn't happen all the time.

I haven't read the code yet but maybe it is related to https://github.com/JuliaLang/julia/issues/11304 ? From `code_warntype` output above (https://github.com/JuliaLang/julia/issues/5117#issuecomment-109292910) it seems that none of the input are bittypes, which currently make the function specialize to a jlcall signature and that causes allocation.

**Comment by yuyichao** [comment]
That seems to what happened.

```julia
julia> s1 = sub(Float64[1, 2], 1:2)
2-element SubArray{Float64,1,Array{Float64,1},Tuple{UnitRange{Int64}},1}:
 1.0
 2.0

julia> @code_warntype getindex(s1, 1)
Variables:
  V::SubArray{Float64,1,Array{Float64,1},Tuple{UnitRange{Int64}},1}
  I::Tuple{Int64}

Body:
  begin 
      $(Expr(:meta, :inline)) # subarray2.jl, line 4:
      checkbounds(V::SubArray{Float64,1,Array{Float64,1},Tuple{UnitRange{Int64}},1},(top(getfield))(I::Tuple{Int64},1)::Int64)::Union(Bool,Void)
      return arrayref((top(getfield))(V::SubArray{Float64,1,Array{Float64,1},Tuple{UnitRange{Int64}},1},:parent)::Array{Float64,1},box(Int64,sub_int(box(Int64,add_int((top(getfield))(V::SubArray{Float64,1,Array{Float64,1},Tuple{UnitRange{Int64}},1},:first_index)::Int64,(top(getfield))(I::Tuple{Int64},1)::Int64)),1)))::Float64
  end::Float64

julia> @code_llvm getindex(s1, 1)
<long llvm IR>
  %16 = call %jl_value_t* @julia_checkbounds_21054(%jl_value_t* inttoptr (i64 139869407299696 to %jl_value_t*), %jl_value_t** %9, i32 2)
<some other llvm IR>
```

And `checkbounds` is specialized to a jlcall signature. (`jl_value_t (*)(jl_value_t*, jl_value_t**, int)`)

**Comment by mbauman** [comment]
Ah, very interesting.  Thanks for popping in here, @yuyichao.  You're right, the type-stability was a non-sequitor.  Making `checkbounds` always return `nothing` actually makes this a bit worse.  I'll have to dig into #11304 more, but do you know why this would have changed?  I should have run the perf tests after rebasing #10525, but performance is *way* worse than I had been seeing a few weeks ago across the board. :-\

@tkelman `./julia test/perf/array/perf.jl` does indeed show this regression.

**Comment by yuyichao** [comment]
I'm leaving now but you can give https://github.com/JuliaLang/julia/pull/11306/files a shot.

**Comment by mbauman** [comment]
Very interesting.  An older branch from a few days ago does not show this performance regression: https://github.com/JuliaLang/julia/compare/JuliaLang:master...mbauman:mb/abstractsmarts11

I'm not sure if something else happened in the few days between the merge commit there and now, or if I goofed in conflict resolution.

**Comment by mbauman** [comment]
The culprit seems to be c8b5c0a.  Removing that from master brings my perf tests back to within their expected range.

**Comment by timholy** [comment]
> but with nightly access to b[-10] triggers an exception due to checkbounds

You can call `Base.sub_unsafe` directly to bypass the checkbounds.

**Comment by timholy** [comment]
> One culprit seems to be c8b5c0a

Surprising culprit. But since that seems to be a correct fix for a type-system bug, we will need to find a way to code this that doesn't depend on the bug.

Unfortunately I'm about to be out-of-touch for the rest of the day, but I can help tomorrow if this isn't resolved.

**Comment by mbauman** [comment]
Since this is more general than just SubArray and I wanted a scratchpad to document my tests, I've moved this to #11595.  Thank you so much for noticing this and bringing it to my attention, @alsam!

**Comment by alsam** [comment]
>> but with nightly access to b[-10] triggers an exception due to checkbounds

> You can call Base.sub_unsafe directly to bypass the checkbounds.

Thanks @timholy for the hint.
Unfortunately it doesn't work for me:
```julia
julia> a = [-10:10];
WARNING: [a] concatenation is deprecated; use collect(a) instead
[...]
julia> b = Base.sub_unsafe(a, 12:21);
julia> b[-10];
ERROR: BoundsError: attempt to access 0-dimensional SubArray{Int64,0,Array{Int64,1},Tuple{Int64},1}:
1
  at index [-10]
 in throw_boundserror at abstractarray.jl:146
 in checkbounds at abstractarray.jl:166
 in getindex at subarray2.jl:4
```
Julia version:
```sh
Version 0.4.0-dev+5236 (2015-06-06 15:02 UTC)
Commit 01abdc3* (0 days old master)
```

``` julia
julia> @code_llvm b[-10]
...
  %10 = icmp eq i32 %2, 1
  br i1 %10, label %fail, label %pass2

fail:                                             ; preds = %top
  %11 = getelementptr %jl_value_t** %1, i64 1
  call void @jl_bounds_error_tuple_int(%jl_value_t** %11, i64 0, i64 1)
  unreachable
...
```
So even `sub_unsafe` checks bounds.

**Comment by alsam** [comment]
Sorry, pressed the wrong button and closed the issue.
Thanks a lot for your efforts!

**Comment by KristofferC** [comment]
This seems to do what worked in 0.3.9.

```julia
julia> a = [-10:10];

julia> b = sub(a, 12:21);

julia> Base.unsafe_getindex(b, -10)
-10
```

**Comment by alsam** [comment]
Sure, it works in 0.3.9, even conventional `b[-10]`.
```julia
julia> a = [-10:10];

julia> b = sub(a, 12:21);

julia> b[-10]
-10

julia> b[-11]
ERROR: BoundsError()
 in getindex at subarray.jl:194
```
Thanks @KristofferC for noticing it.

**Comment by yuyichao** [comment]
@alsam I think you need `unsafe_getindex` on `0.4`

```julia
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "help()" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.0-dev+5236 (2015-06-06 15:02 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 01abdc3* (1 day old master)
|__/                   |  x86_64-unknown-linux-gnu

julia> a = [-10:10;];

julia> b = sub(a, 12:21);

julia> Base.unsafe_getindex(b, -10)
-10
```

**Comment by alsam** [comment]
Copy that.
```
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "help()" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.0-dev+5236 (2015-06-06 15:02 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 01abdc3* (1 day old master)
|__/                   |  x86_64-linux-gnu

julia> a = [-10:10;];

julia> b = sub(a, 12:21);

julia> Base.unsafe_getindex(b, -10)
-10

julia> Base.unsafe_getindex(b, -11)
ERROR: BoundsError: attempt to access 21-element Array{Int64,1}:
 -10
...
```
Thanks @yuyichao for the useful hint!

**Comment by KristofferC** [comment]
@alsam I meant that what I wrote does on 0.4 what used to work on 0.3.9. In fact, @yuyichao posted the exact same code as me :P

**Comment by yuyichao** [comment]
> In fact, @yuyichao posted the exact same code as me :P

:) and that's exactly what I did in my julia session. Just want to include the version info and point out that it is not `unsafe_sub` rather `unsafe_getindex`.

**Comment by alsam** [comment]
Thank you @KristofferC and @yuyichao for clarification.
Got it :)
Just curious, what is the best way to wrap `unsafe_getindex` to get convenient short sugared syntax `b[-10]` instead of lengthy `unsafe_getindex`? Have some ideas in mind but maybe you might propose better way?

**Comment by KristofferC** [comment]
The code below is probably fit for an insanity wolf meme hehe.

```julia
julia> import Base: getindex

julia> const getindex = Base.unsafe_getindex
Warning: imported binding for getindex overwritten in module Main
unsafe_getindex (generic function with 19 methods)

julia> a = [-10:10;];

julia> b = sub(a, 12:21);

julia> b[-10]
-10
```

But seriously, you should probably be able to do the same but only for the `SubArray` `getindex` methods.

**Comment by alsam** [comment]
Excellent, thanks @KristofferC !

**Comment by KristofferC** [comment]
Note, you should do something like
```julia
getindex(V::SubArray, I::Int64...) = Base.unsafe_getindex(V, I...)
getindex(V::SubArray, I::Union(Range{Int64},Array{Int64,1},Int64,Colon)...) = Base.unsafe_index(V, I...)
getindex{T,N,P,IV,LD}(V::SubArray{T,N,P,IV,LD}, I::Union(AbstractArray{T,N},Colon,Real)...) = Base.unsafe_index(V, I...)
```

instead to keep the boundschecking for non subarray arrays.

**Comment by alsam** [comment]
Cool, thanks a lot @KristofferC :))
Just checked, worked as expected:
```julia
julia> b[-10]
-10

julia> b[-11]
ERROR: BoundsError: attempt to access 21-element Array{Int64,1}:
```

**Comment by timholy** [comment]
Chiming in late here, but yes, indexing changed a lot for #10525, so my old instincts are out of date. Glad y'all figured stuff out.

**Comment by alsam** [comment]
Let me summarize the performance, with the following code below - thanks to @KristofferC, @timholy,  @yuyichao for the hints
```julia
getindex(V::SubArray, I::Int...) = Base.unsafe_getindex(V, I...)
getindex(V::SubArray, I::Union(Range{Int},Array{Int,1},Int,Colon)...) = Base.unsafe_getindex(V, I...)
getindex{T,N,P,IV}(V::SubArray{T,N,P,IV}, I::Union(Real, AbstractArray, Colon)...) = Base.unsafe_getindex(V, to_index(I)...)
getindex{T,N,P,IV,LD}(V::SubArray{T,N,P,IV,LD}, I::Union(AbstractArray{T,N},Colon,Real)...) = Base.unsafe_getindex(V, I...)
setindex!(V::SubArray, v, I::Int...) = Base.unsafe_setindex!(V, v, I...)


#import Base: getindex, setindex!
#const getindex = Base.unsafe_getindex
#const setindex! = Base.unsafe_setindex!
```
the performance is:
```sh
follow up: mykern( 2:(siz - 3), s, a, x, y )
  12.896 milliseconds (166 allocations: 11485 bytes)
follow up: mykern( 2:(siz - 3), s1, a, x1, y1 )
  81.872 milliseconds (16694 allocations: 713 KB)
follow up: mykern( 3:(siz - 2), s2, a, x2, y2 )
  19.551 milliseconds (6 allocations: 192 bytes)
  19.710 milliseconds (6 allocations: 192 bytes)
  19.322 milliseconds (6 allocations: 192 bytes)
```
that is the performance penalty is about 40% as @timholy predicted.
That is excellent. Interesting that for 0.3.9 didn't observe performance penalty at all for `sub` indexing:
```sh
follow up: 
mykern( 2:(siz - 3), s, a, x, y )
elapsed time: 0.013599747 seconds (13888 bytes allocated)
follow up: mykern( 2:(siz - 3), s1, a, x1, y1 )
elapsed time: 0.170130061 seconds (938696 bytes allocated)
follow up: mykern( 3:(siz - 2), s2, a, x2, y2 )
elapsed time: 0.124578779 seconds (88 bytes allocated)
elapsed time: 0.114813522 seconds (88 bytes allocated)
elapsed time: 0.113371351 seconds (88 bytes allocated)
```

**Comment by timholy** [comment]
@alsam, you're still not noticing that 0.3.9 has a **factor of ten** performance penalty for `sub` indexing. (`0.12... seconds` is 120 milliseconds, not 12 millseconds).

#11617 makes it a lot better (and better than on 0.3), but it's still not back to where we were a couple weeks ago.

**Comment by alsam** [comment]
Yes, @timholy , you are quite right, thanks for noticing and for the link #11617 .

**Comment by alsam** [comment]
Ok, let me summarize:
Just a few lines of code emulate Fortran-like arrays quite effectively:
```julia
getindex(V::SubArray, I::Int64...) = Base.unsafe_getindex(V, I...)
getindex(V::SubArray, I::Union(Range{Int64},Array{Int64,1},Int64,Colon)...) = Base.unsafe_getindex(V, I...)
getindex{T,N,P,IV}(V::SubArray{T,N,P,IV}, I::Union(Real, AbstractArray, Colon)...) = Base.unsafe_getindex(V, to_index(I)...)
setindex!(V::SubArray, v, I::Int64...) = Base.unsafe_setindex!(V, v, I...)

macro shifted_array(T, r)
    :(sub(Array($T, length($r)), -minimum($r) + 2 : length($r)))
end
```
and usage:
```julia
julia> x = @shifted_array(Float64, -10:10)
10-element SubArray{Float64,1,Array{Float64,1},Tuple{UnitRange{Int64}},1}:
 6.90782e-310
...
julia> x[-10]
6.9078200935909e-310

julia> x[-11]
ERROR: BoundsError: attempt to access 21-element Array{Float64,1}:
...
julia> x[0]
6.90782011166424e-310

julia> x[10]
6.9078200930178e-310

julia> x[11]
ERROR: BoundsError: attempt to access 21-element Array{Float64,1}:
```
Penalty for a bigger example from 1d acoustics is quite reasonable - ~40% as expected:
timings, for `base` line:
```sh
593.164 milliseconds (40 allocations: 236 KB)
```
for `sub` version that uses macro `@shifted_array`:
```
 874.729 milliseconds (19507 allocations: 540 KB)
```

**Comment by alsam** [comment]
It is fantastic, but with the latest Julia `Version 0.4.0-dev+5293 (2015-06-10 09:21 UTC)` and `@inline`-ed `getindex/setindex!`:
```julia
@inline getindex(V::SubArray, I::Int64...) = Base.unsafe_getindex(V, I...)
@inline getindex(V::SubArray, I::Union(Range{Int64},Array{Int64,1},Int64,Colon)...) = Base.unsafe_getindex(V, I...)
@inline getindex{T,N,P,IV}(V::SubArray{T,N,P,IV}, I::Union(Real, AbstractVector, Colon)...) = Base.unsafe_getindex(V, to_index(I)...)
@inline setindex!(V::SubArray, v, I::Int64...) = Base.unsafe_setindex!(V, v, I...)
```
(please note, that `@inline getindex{T,N,P,IV}(V::SubArray{T,N,P,IV}, I::Union(Real, AbstractVector ...` follow https://github.com/JuliaLang/julia/commit/c7e2b33030bb039e05af4ced931da07163fc4dc5 .)
`sub` version works faster in some cases:
```sh
julia main_sub.jl
...
333.805 milliseconds (19501 allocations: 539 KB)
```
vs.
```sh
julia main.jl
...
633.032 milliseconds (40 allocations: 236 KB)
```
but this is due to `unsafe_getindex/unsafe_setindex!`
if we define
```julia
import Base: getindex, setindex!
const getindex = Base.unsafe_getindex
const setindex! = Base.unsafe_setindex!
```
then we will get 
```sh
julia main.jl 
Warning: imported binding for getindex overwritten in module Main
Warning: imported binding for setindex! overwritten in module Main
...
 198.278 milliseconds (40 allocations: 236 KB)
```
as expected, i.e. regular array version is faster than `sub` one. As byproduct check bounds is quite costly, without it the regular array version is >3x faster for this particular case.

**Comment by mbauman** [comment]
I've seen some strange performance effects due to using `Base.f()` in function definitions.  I never fully tracked it down, but perhaps try instead:

```jl
import Base: getindex, unsafe_getindex
@inline getindex(V::SubArray, I::Int64...) = unsafe_getindex(V, I…)
```

We're planning on reverting back to the old SubArray semantics, where bounds checks are deferred to the parent array (which is then affected by `@inbounds`).  See #11641.

**Comment by alsam** [comment]
Great, it would be nice to have old SubArray semantics reverted, thanks @mbauman !

**Comment by alsam** [comment]
Confirm that with the last julia `Version 0.4.0-dev+5394 (2015-06-15 23:17 UTC)`  even without redefining `@inline getindex(V::SubArray, I::Int64...) = unsafe_getindex(V, I...)` the `sub` trick to achieve negative indexing works out of the box and checks bounds correctly:
```julia
julia> macro shifted_array(T, r)
           :(sub(Array($T, length($r)), -minimum($r) + 2 : length($r)))
       end

julia> z = @shifted_array(Float64, -7:7);

julia> z[-7]
6.92464944583943e-310

julia> z[-8]
ERROR: BoundsError: attempt to access 15-element Array{Float64,1}:
 6.92465e-310
```
The last score for 1d acoustics code

baseline:
```sh
 320.611 milliseconds (40 allocations: 236 KB)
```
vs. `sub` version:
```sh
483.243 milliseconds (19501 allocations: 539 KB)
```
so the penalty is quite reasonable.
Thanks @mbauman, @timholy, @KristofferC, @yuyichao !
